contracts/0.8.9/oracle/ValidatorsExitBusOracle.sol

```diff
@@ -0,0 +1,507 @@
+        refSlot = _storageConsensusReport().value.refSlot;
+        processingStarted = state.refSlot != 0 && state.refSlot == refSlot;
+        if (processingStarted) {
+            requestsCount = state.requestsCount;
+            requestsProcessed = state.requestsProcessed;
+            dataFormat = state.dataFormat;
+        }
+    }
+
+    ///
+    /// Implementation & helpers
+    ///
+
+    function _setDataBoundaries(
+        uint256 maxRequestsPerReport,
+        uint256 maxRequestsListLength,
+        uint256 rateLimitWindowSlots,
+        uint256 rateLimitMaxThroughputE18
+    ) internal {
+        uint256 currentRefSlot = _getCurrentRefSlot();
+
+        _storageDataBoundaries().value = DataBoundaries({
+            maxRequestsPerReport: maxRequestsPerReport.toUint64(),
+            maxRequestsListLength: maxRequestsListLength.toUint64()
+        });
+
+        RateLimit
+            .load(RATE_LIMIT_POSITION)
+            .configureThroughput(currentRefSlot, rateLimitWindowSlots, rateLimitMaxThroughputE18)
+            .store(RATE_LIMIT_POSITION);
+
+        emit DataBoundariesSet(
+            currentRefSlot,
+            maxRequestsPerReport,
+            maxRequestsListLength,
+            rateLimitWindowSlots,
+            rateLimitMaxThroughputE18
+        );
+    }
+
+    function _handleConsensusReport(
+        ConsensusReport memory /* report */,
+        uint256 /* prevSubmittedRefSlot */,
+        uint256 prevProcessingRefSlot
+    ) internal override {
+        DataProcessingState memory state = _storageDataProcessingState().value;
+        if (state.refSlot == prevProcessingRefSlot && state.requestsProcessed < state.requestsCount) {
+            emit WarnDataIncompleteProcessing(
+                prevProcessingRefSlot,
+                state.requestsProcessed,
+                state.requestsCount
+            );
+        }
+    }
+
+    function _checkMsgSenderIsAllowedToSubmitData() internal view {
+        address sender = _msgSender();
+        if (!hasRole(SUBMIT_DATA_ROLE, sender) && !_isConsensusMember(sender)) {
+            revert SenderNotAllowed();
+        }
+    }
+
+    function _handleConsensusReportData(ReportData calldata data) internal {
+        if (data.dataFormat != DATA_FORMAT_LIST) {
+            revert UnsupportedRequestsDataFormat(data.dataFormat);
+        }
+
+        if (data.data.length % 64 != 0) {
+            revert InvalidRequestsDataLength();
+        }
+
+        if (data.data.length / 64 != data.requestsCount) {
+            revert UnexpectedRequestsDataLength();
+        }
+
+        uint256 lastProcessedItemWithoutPubkey = _processExitRequestsList(data.data);
+
+        _storageDataProcessingState().value = DataProcessingState({
+            lastProcessedItemWithoutPubkey: lastProcessedItemWithoutPubkey,
+            refSlot: data.refSlot.toUint64(),
+            requestsCount: data.requestsCount.toUint64(),
+            requestsProcessed: data.requestsCount.toUint64(),
```

Comment:
Why do we need two separate items `requestsCount ` and `requestsProcessed`

Reply:
since exit requests can be brought in multiple transactions after Merkle data format is implemented